### PR TITLE
WIP: Attendance Event, Event User Values, and Events Rework

### PR DESCRIPTION
**Pull request by Zintixx**
- This implements (and activates) attendance event. I'm only saying WIP because a lot of things may change.
 - It is fully functional but however, the backend of things likely need to be shifted around, particularly Event User Values. With this being the first event using event user values as of now, things may be entirely reworked once more to get a better understanding on how they're used in a more global scale.
 - Moved over updating a player's log time out of the OnlineTimer loop in order for it to be of use for AttendanceEvent.
 - temporarily hardcoding a attendance mail. This likely needs something more flexible to support more custom mails

### CODE REVIEW

File: MapleServer2/Types/GameEventUserValue.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+using MapleServer2.Enums;
+
+namespace MapleServer2.Types;
+
+public class GameEventUserValue
+{
+    public long Id;
+    public long CharacterId;
+    public GameEventUserValueType EventType;
+    public string EventValue;
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/GameEventUserValue.cs:11]: So this value is always a number but in the packet is always a string? If that's the case I would make this a int and only use ToString() on the packet.

File: MapleServer2/Types/TimeInfo.cs [hunk 2]
```diff
@@ -7,4 +7,8 @@ public static class TimeInfo
     public static DateTime CurrentDate() => DateTimeOffset.UtcNow.UtcDateTime;
 
     public static long AddDays(int days) => DateTimeOffset.UtcNow.AddDays(days).ToUnixTimeSeconds();
+
+    public static long Tomorrow() => (long) DateTime.Now.AddDays(1).ToUniversalTime().Date.Subtract(new DateTime(1970, 1, 1)).TotalSeconds;
+
+    public static DateTime TimestampToDateTime(long timestamp) => new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc).AddSeconds(timestamp);
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/TimeInfo.cs:13]: Can't you just parse the timestamp???

File: MapleServer2/Types/TimeInfo.cs [hunk 3]
```diff
@@ -7,4 +7,8 @@ public static class TimeInfo
     public static DateTime CurrentDate() => DateTimeOffset.UtcNow.UtcDateTime;
 
     public static long AddDays(int days) => DateTimeOffset.UtcNow.AddDays(days).ToUnixTimeSeconds();
+
+    public static long Tomorrow() => (long) DateTime.Now.AddDays(1).ToUniversalTime().Date.Subtract(new DateTime(1970, 1, 1)).TotalSeconds;
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/TimeInfo.cs:11]: Why substract 0?

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: MapleServer2/Types/TimeInfo.cs (hunk 3 above)
Comment (AngeloTadeucci) [reply MapleServer2/Types/TimeInfo.cs:11]:
```csharp
        ((DateTimeOffset) DateTime.Now.AddDays(1).Date).ToUnixTimeSeconds(); // without timezone, without time
        ((DateTimeOffset) DateTime.Now.AddDays(1).ToUniversalTime().Date).ToUnixTimeSeconds(); // UTC, without time
        DateTimeOffset.UtcNow.AddDays(1).ToUnixTimeSeconds(); // utc, with time
```

File: MapleServer2/Types/TimeInfo.cs (hunk 2 above)
Comment (AngeloTadeucci) [reply MapleServer2/Types/TimeInfo.cs:13]: `DateTimeOffset.FromUnixTimeSeconds(unixTimestamp)`

File: MapleServer2/Types/GameEventUserValue.cs (hunk 1 above)
Comment (Zintixx) [reply MapleServer2/Types/GameEventUserValue.cs:11]: not always an int. It's sometimes a bool, int, or long

**Review by AngeloTadeucci — APPROVED** [review]

**Pull request merged**